src/Compilers/Core/Portable/Symbols/ITypeSymbol.cs

```diff
@@ -125,6 +135,15 @@ public interface ITypeSymbol : INamespaceOrTypeSymbol
         /// <returns>A formatted string representation of the symbol with the given nullability.</returns>
         string ToDisplayString(NullableFlowState topLevelNullability, SymbolDisplayFormat format = null);
 
+        /// <summary>
+        /// Converts a symbol to an array of string parts, each of which has a kind. Useful
+        /// for colorizing the display string.
+        /// </summary>
+        /// <param name="topLevelNullability">The top-level nullability to use for formatting.</param>
+        /// <param name="format">Format or null for the default.</param>
+        /// <returns>A read-only array of string parts.</returns>
+        ImmutableArray<SymbolDisplayPart> ToDisplayParts(NullableAnnotation topLevelNullability, SymbolDisplayFormat format = null);
```

Comment:
>ToDisplayParts [](start = 42, length = 14)

Please add some compiler-level tests for those new compiler APIs, on C# and VB sides.

Reply:
We don't have any tests for the existing new APIs, and they will be refactored. I think it's fine for now.

Reply:
Let's sync tomorrow.

Reply:
Didn't get to sync yesterday due to misc urgent issues. 
In short, I think we should add tests. That's the standard we apply for compiler APIs. 
Existing test gaps do not seem relevant to me (assuming we think we have an existing mistake, we shouldn't do more of it).
If there is an issue of significant cost, timing or churn (API expected to change), we should discuss some options (branch, do basic testing and defer rest with tracking issue, ...).

Reply:
@333fred Is there an existing tracking issue for this?

I think it's fair to say that right now there's at least strong discussions of this API changing; even if we don't do the merged ITypeSymbol we're suspecting the NullableAnnotation/NullableFlowState enum split is going to be problematic for the IDE.

I'm happy to add tests, but I'm not sure if your caveats there actually _do_ apply in this case.

Reply:
There's a tracking issue for refactoring the API implementations here: https://github.com/dotnet/roslyn/issues/35035

Reply:
So am I good to continue with this without tests? @jcouv? I don't care which way we go as much as a clear direction.

Reply:
Let's add tests. It shouldn't be hard and I think this API will remain in some shape or another.
Thanks

For display string:
We have one formatting option that is relevant, and it can be on or off. 
The "off" option can be tested with `string` and four annotations (`NotApplicable`, `Disabled`, `NotAnnotated` and `Annotated`).
The "on" option is more interesting, so I would cover the matrix of three cases (`TClass`, `TStruct` and `TOpen`) and four annotations. Then maybe one generic case (`List<string?>` with annotation) and an invalid case (`object*` with annotation).

For display parts: I think we'll get most coverage from display string, so I'm fine with spot checking one scenario: `string` with annotation.

Reply:
API is no longer present now that https://github.com/dotnet/roslyn/pull/39498 is merged.